content/blog/_posts/2021-04-rust-custom-test-harness.md

```diff
@@ -0,0 +1,296 @@
+---
+title: How to Build a Custom Test Harness in Rust
+author:
+    name: "T.J. Telan"
+    github: "tjtelan"
+description: "Building a custom test harness in Rust is less complicated following these patterns"
+date: 2021-04-26
+slug: rust-custom-test-harness 
+url: /blog/2021/04/rust-custom-test-harness
+img: blog/images/rust-custom-harness/test-harness-social.png
+tile: blog/images/rust-custom-harness/test-harness-social.png
+twitter-card: summary_large_image
+code:
+    height: 9000
+---
+
+## Let’s talk about integration testing in Rust
+
+I ran into a problem effectively using `cargo test` in [Fluvio](https://github.com/infinyon/fluvio) for integration testing. But I learned that you aren’t stuck with the functionality of `cargo test`.
+
+In this post, we’ll go implementing a custom test harness and some of the decision making that led me to this type of solution.
+
+
+### How does cargo test work by default?
+
+There is a distinction between unit tests and integration tests in [the Rust book](https://doc.rust-lang.org/book/ch11-03-test-organization.html). The distinction is less about testing strategy and more about defining Rust’s conventions for test organization.
+
+The main points are that:
+  * Your tests are annotated with `#[test]`
+  * [libtest](https://github.com/rust-lang/libtest) harness enumerates through all of your tests (a point we’ll revisit later in more detail)
+  * libtest returns the pass/fail status of the execution
+
+
+### What do I need from integration testing?
+
+There isn’t anything libtest specifically offers to support integration testing patterns.
+
+Setup of a standard test environment, especially in a complex system, is essential for managing expected behavior when making code changes.
+
+Libtest does not assist with setup or teardown. I need the ability to abstract away the setup and teardown of my test environment from test code. 
+
+This task will be performed either way. Without harness support, setup/teardown will be performed via external shell scripts or padding the setup/teardown process within every single integration test...
+
+It isn’t convenient to manage setup and teardown in a different context than the integration test. This kind of testing overhead leads to hard to reproduce and time consuming mistakes.
+
+### Where do we get started with a custom test harness?
+
+We’re going to build all our integration tests into a single crate. This is recommended in order to speed up compile time ([[1](https://endler.dev/2020/rust-compile-times/#combine-all-integration-tests-in-a-single-binary)], [[2](https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html)])
```

Comment:
Try using this way of defining links in order to make the text representation more readable:

```
We’re going to build all our integration tests into a single crate. This is recommended in order to speed up compile time ([[1]], [[2]])

[1]: https://endler.dev/2020/rust-compile-times/#combine-all-integration-tests-in-a-single-binary
[2]: https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html
```